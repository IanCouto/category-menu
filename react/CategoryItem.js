import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'
import classNames from 'classnames'

import categoryQuery from './queries/categoryQuery.gql'

/**
 * Component that represents a single category displayed in the menu, also displays
 * the subcategories, if the provided category has them
 */
class CategoryItem extends Component {
  render() {
    const { data: { category } } = this.props

    const wrapperClasses = classNames('h3 pl5 pr4 vtex-category-item', {
      'show-arrow': category.hasChildren,
    })

    return (
      <div className={wrapperClasses}>
        <a href={category.href} className="db mt6 no-underline ttu black-90">
          {category.name}
        </a>

        {category.hasChildren && (
          <div className="vtex-category-sub-menu pv6 ph5 br2 br--bottom">
            <ul className="list ma0 pa0 f6">
              {category.children.map(subCategory => (
                <li key={subCategory.id} className="lh-copy">
                  <a
                    className="near-black no-underline underline-hover"
                    href={subCategory.href}
                  >
                    {subCategory.name}
                  </a>
                </li>
              ))}
            </ul>
          </div>
        )}
      </div>
    )
  }
}

CategoryItem.propTypes = {
  data: PropTypes.shape({
    category: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      href: PropTypes.string.isRequired,
      hasChildren: PropTypes.bool.isRequired,
    }),
    loading: PropTypes.bool.isRequired,
  }).isRequired,
  /** Id of the category */
  id: PropTypes.number.isRequired,
}

const options = {
  options: props => ({
    variables: {
      id: props.id,
    },
  }),
}

export default graphql(categoryQuery, options)(CategoryItem)
