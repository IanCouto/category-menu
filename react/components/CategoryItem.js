import React, { Component } from 'react'
import { Link } from 'vtex.render-runtime'
import PropTypes from 'prop-types'
import { categoryItemShape } from '../propTypes'

import ItemContainer from './ItemContainer'
import classNames from 'classnames'
import categoryMenu from '../categoryMenu.css'


/**
 * Component that represents a single category displayed in the menu, also displays
 * the subcategories, if the provided category has them
 */
export default class CategoryItem extends Component {
  state = {
    isOnHover: false,
  }

  static propTypes = {
    /** Category to be displayed */
    category: categoryItemShape.isRequired,
    /** Set use of Link component */
    noRedirect: PropTypes.bool,
    /** Number of subcategory levels */
    subcategoryLevels: PropTypes.oneOf([0, 1, 2]),
  }

  handleCloseMenu = () => (this.setState({ isOnHover: false }))

  renderCategory(){
    const { category, noRedirect } = this.props
    const { isOnHover } = this.state

    const linkClasses = classNames(
      'w-100 pv5 mh6 no-underline t-small outline-0 db tc link truncate bb bw1 c-muted-1', {
        'b--transparent': !isOnHover,
        'b--action-primary': isOnHover,
      }
    )

    return noRedirect ? (
      <a href="#" className={linkClasses}>
        {category.name.toUpperCase()}
      </a>
    ) : (
      <Link
        onClick={this.handleCloseMenu}
        page="store.search#department"
        params={{ department: category.slug }}
        className={linkClasses}
      >
        {category.name.toUpperCase()}
      </Link>
    )
  }

  renderChildren(){
    const { category, subcategoryLevels } = this.props
    const { isOnHover } = this.state

    const containerStyle = {
      top: this.item && this.item.offsetTop + this.item.clientHeight,
      display: isOnHover ? 'flex' : 'none',
    }

    return subcategoryLevels > 0 && category.children.length > 0 && (
      <ItemContainer
        containerStyle={containerStyle}
        categories={category.children}
        parentSlug={category.slug}
        onCloseMenu={this.handleCloseMenu}
        showSecondLevel={subcategoryLevels === 2}
      />
    )
  }
  
  render() {
    return (
      <li className={`${categoryMenu.container, categoryMenu.list} flex items-center db list`}
        ref={e => { this.item = e }}
        onMouseEnter={() => this.setState({ isOnHover: true })}
        onMouseLeave={this.handleCloseMenu}
      >
        {this.renderCategory()}
        {this.renderChildren()}
      </li>
    )
  }
}
