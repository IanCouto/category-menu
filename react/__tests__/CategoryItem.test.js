import React from 'react'
import { mount } from 'enzyme'

import CategoryItem from '../components/CategoryItem'

describe('CategoryItem component', () => {
  let wrapper

  beforeEach(() => {
    const categoryMock = {
      id: 1,
      name: 'Category',
      href: '#1',
      slug: 'category',
      hasChildren: true,
      children: [
        {
          id: 2,
          name: 'Sub-category',
          href: '#2',
          slug: 'sub-category',
        },
      ],
    }

    wrapper = mount(<CategoryItem category={categoryMock} />)
  })

  it('should have the correct href', () => {
    expect(wrapper.getDOMNode().querySelector('a').href).toBe(
      `${location.href}#1`
    )
  })

  it('should match snapshot', () => {
    expect(wrapper.getElement()).toMatchSnapshot()
  })
})
