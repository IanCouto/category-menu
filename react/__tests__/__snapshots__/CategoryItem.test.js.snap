// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem component should match snapshot 1`] = `
<div
  className="vtex-category-menu__item flex justify-center items-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Link
    className="w-100 pv5 mh6 no-underline t-action--small outline-0 db tc truncate bb bw1 c-muted-1 b--transparent"
    onClick={[Function]}
    page="store/department"
    params={
      Object {
        "department": "category",
      }
    }
  >
    CATEGORY
  </Link>
</div>
`;
